### PR TITLE
IDE friendly errors

### PR DESCRIPTION
**Pull request by alloy**
These changes make it so file/line locations in errors can be made linkable by IDEs. This is the format used by jest, flow, typescript, and many others.

### Before

relay-ide-friendly-errors-before

### After

relay-ide-friendly-errors-after

### CODE REVIEW

File: packages/relay-compiler/core/RelayCompilerError.js [hunk 1]
```diff
@@ -110,7 +110,7 @@ function createCombinedError(
     `${prefix}Encountered ${errors.length} error(s):\n` +
       errors
         .map(error =>
-          String(error)
+          (error instanceof GraphQLError ? printError(error) : String(error))
```
Comment (alloy) [inline packages/relay-compiler/core/RelayCompilerError.js]: Use our own formatter for `GraphQLError` instances, instead of the one from graphql-js.

File: packages/relay-compiler/core/RelayCompilerError.js [hunk 2]
```diff
@@ -192,6 +192,34 @@ function printLocations(locations: $ReadOnlyArray<Location>): Array<string> {
...
+  } else if (error.source && error.locations) {
+    const source = error.source;
+    for (const location of error.locations) {
+      printedLocations.push(highlightSourceAtLocation(source, location));
+    }
+  }
+  return printedLocations.length === 0
+    ? error.message
+    : [error.message, ...printedLocations].join('\n\n') + '\n';
+}
```
Comment (alloy) [inline packages/relay-compiler/core/RelayCompilerError.js]: All only needed so we can use our own copy of `highlightSourceAtLocation`.

File: packages/relay-compiler/core/RelayFindGraphQLTags.js [hunk 3]
```diff
@@ -23,14 +23,14 @@ import type {
   GraphQLTagFinder,
 } from '../language/RelayLanguagePluginInterface';
 
-const cache = new RelayCompilerCache('RelayFindGraphQLTags', 'v1');
+const cache = new RelayCompilerCache('RelayFindGraphQLTags', 'v2');
```
Comment (alloy) [inline packages/relay-compiler/core/RelayFindGraphQLTags.js]: Have to bust the cache, as we should now cache the entire `tag`, instead of only the `tag.template`.

File: packages/relay-compiler/core/RelaySourceModuleParser.js [hunk 4]
```diff
@@ -66,14 +66,18 @@ module.exports = (
 
     const astDefinitions = [];
     const sources = [];
-    memoizedTagFinder(text, baseDir, file).forEach(template => {
-      const source = new GraphQL.Source(template, file.relPath);
+    memoizedTagFinder(text, baseDir, file).forEach(tag => {
+      const source = new GraphQL.Source(
+        tag.template,
+        path.join(path.relative(process.cwd(), baseDir), file.relPath),
```
Comment (alloy) [inline packages/relay-compiler/core/RelaySourceModuleParser.js]: Keep path short by making it relative, but in a way that the IDE can find it (i.e. relative to the working dir).

**Comment by alloy** [comment]
It strikes me that all of the code in https://github.com/facebook/relay/blob/c83b7d3408d6938d54b6973d4fe788d3b636e44a/packages/relay-compiler/core/RelayCompilerError.js that’s imported from https://github.com/graphql/graphql-js/blob/c3dd67050eaf8a62fd4907e155e99c6a56bb7066/src/error/printError.js only really is imported because it isn’t exported upstream. I would rather have these file/line locations formatted correctly for IDEs anywhere graphql-js is used, so would you be open to my trying to get those exported upstream and remove them from our codebase?

**Comment by nodkz** [comment]
Yay! ❤❤❤

**Comment by alloy** [comment]
Unsure why the ‘push’ build failed, but it’s due to some git branch issue.

**Comment by alunyov** [comment]
@alloy thank you for this PR. The demo looks super awesome!

We may need more time to merge this internally, because this PRs, as is, is breaking our internal compiler setup. 

I'll follow up on the progress.

**Comment by IvanGoncharov** [comment]
> It strikes me that all of the code in RelayCompilerError that’s imported from graphql-js is only done because it isn’t exported upstream. I would rather have these file/line locations formatted correctly for IDEs anywhere graphql-js is used, so would you be open to my trying to get those exported upstream and remove them from our codebase?

@alloy Thanks for pinging me on Slack. I merged https://github.com/graphql/graphql-js/pull/1984 that expose new `printLocation` and `printSourceLocation` functions.
You can try it on using `npm` branch: https://github.com/graphql/graphql-js#want-to-ride-the-bleeding-edge
Would be great if you rebase your PR on top of this branch so we can figure out if `printSourceLocation` location will work for Relay.

**Comment by alloy** [comment]
@IvanGoncharov Great stuff, thanks! I’ll give it a try and let you know 👍

**Comment by alloy** [comment]
@IvanGoncharov Follow-up is at https://github.com/facebook/relay/pull/2784

**Comment by kassens** [comment]
@alloy Sorry for the delay here, can you summarize the latest status? Is this PR made redundant due to the graphql-js change and we just need #2784?


(I'm not a big fan how GitHub handles dependent changes).

**Comment by alloy** [comment]
@kassens No worries–no need ever to apologise for not having time to work on external OSS!

So, semantically speaking this PR is still needed and separate from #2784. However, seeing as those changes _also_ exist in #2784 _and_ if you are ready to bump the version constraint of the `graphql-js` dependency, it might be easier on your end to just merge #2784 separately. My suggestion for future maintenance purposes, however, would be to still merge them separately, as they are imo distinct units of work.

In the meantime I'll rebase these PRs and I'll be available the rest of the day for further coordination on these.

File: packages/relay-compiler/core/RelayFindGraphQLTags.js (hunk 3 above)
Comment (alloy) [reply packages/relay-compiler/core/RelayFindGraphQLTags.js]: @kassens This work is distinct from #2784 imo.

File: packages/relay-compiler/core/RelaySourceModuleParser.js (hunk 4 above)
Comment (alloy) [reply packages/relay-compiler/core/RelaySourceModuleParser.js]: @kassens This too.

**Comment by kassens** [comment]
Since we're 100% focused on the compiler rewrite and this required some internal changes. I'll close this for now. Better IDE integration is one of our new goals :)
